### ISSUE TITLE
Inconsistent escaping behavior

### ISSUE THREAD

**Issue by legutierr** [twigphp/Twig#1997]
A) With auto escaping turned on, double escaping happens sometimes but not other times.  For example:

    >>> $loader = new Twig_Loader_Filesystem(PATH);
    >>> $twig = new Twig_Environment($loader, array('autoescape' => true));
    >>> $twig->createTemplate("{% set var = '<a>' %}{% set val = var|e %}{{ val }}")->render(array())
    => "&amp;lt;a&amp;gt;"
    >>> $twig->createTemplate("{% set var = '<a>' %}{% set val = var %}{{ val|e }}")->render(array())
    => "&lt;a&gt;"

In other words, `{% set val = var|e %}{{ val }}` and `{% set val = var %}{{ val|e }}` produce different results.  This does not seem like correct behavior.

B) This seems related to this double-escaping bug:

    >>> $twig->createTemplate("{% set val = '<a>' %}{% set var = val|e %}{{ var|e }}")->render(array())
    => "&amp;lt;a&amp;gt;"

The escape filter should not escape something that has already been escaped.  The correct output should be `"&lt;a&gt;"`

**Comment by xabbuh** [comment]
I don't think that this is a bug. You cannot reliably determine if a value really is escaped properly.  When you know that your data already is escaped, you can use the `raw` filter to output it without auto escaping taking effect.

**Comment by stof** [comment]
This is because autoescaping checks whether escaping is needed on display based on the syntax of the template. so ``{{ val|e }}`` will see that the displayed value is already escaped (as ``|e`` marks the value as safe when being applied), while ``{{ val }}`` does not and so escaping is needed.

**Comment by legutierr** [comment]
@stof The thing that seems very inconsistent is that `|e` marks the value as safe when applied in the location of `{{ val|e }}`, but `val` doesn't retain the safe "mark" when `|e` is used inside `{% set val = var|e %}`.  Why doesn't `val` get set to "var-as-marked" instead of "var-as-un-marked"?  

With this behavior, it seems as if the filter is acting as some kind of annotation on the expression tag to give some kind of instruction to the autoescape engine, and is not actually operating on the level of the data.  Filters are supposed to operate on data: they accept data and output transformed data, and the only thing influencing the output should be the inputs, as filters are functions.  But here, the complete description of the type of data being output by the filter is influenced by *where* the expression executing.  You would never see this kind of behavior in a programming language.

The issue is more obvious with the `raw` filter:

>>> $twig->createTemplate("{% set val = '<a>'|raw %}{% set var = val %}{{ var|raw }}")->render(array()) 
=> "<a>"
>>> $twig->createTemplate("{% set val = '<a>'|raw %}{% set var = val|raw %}{{ var }}")->render(array()) 
=> "&lt;a&gt;"

What does the `raw` filter do? Does it take as input some string, and output a string marked as safe?  That's what it looks like in the first case.  But in the second case, it's as if it does absolutely nothing!

**Comment by stof** [comment]
@legutierr it does not mark the value itself as safe. It marks the expression as safe. Reusing the variable elsewhere is a different Twig expression.

**Comment by stof** [comment]
All the auto-escaping is compile-time logic to determine whether an escaping should be added or no.

and values themselves cannot be marked as safe during runtime as it would require to wrap values into a special object holding this info, which would make it a huge pain to use Twig extension points and a performance drawback as values would have to be unwrapped very frequently (when passing them to filters, etc...).
Symfony tried to provide auto-escaping in its PHP templating engine with such an approach in the past, and gave up as it was very buggy, hard to use and slow (which is one of the reasons Twig is the recommended engine in Symfony since years).
